lib/go-tc/invalidationjobs.go

```diff
@@ -456,3 +463,210 @@ func (job *UserInvalidationJobInput) Validate(tx *sql.Tx) error {
+	errs := []string{}
+	err := validation.ValidateStruct(job,
+		validation.Field(&job.DeliveryService, validation.Required),
+		validation.Field(&job.AssetURL, validation.Required, is.URL),
+		validation.Field(&job.CreatedBy, validation.Required),
+		validation.Field(&job.ID, validation.Required),
+		validation.Field(&job.TTLHours, validation.Required),
+		validation.Field(&job.InvalidationType, validation.Required, validation.NewStringRule(func(s string) bool {
+			return s == REFRESH || s == REFETCH
+		}, fmt.Sprintf("must be either %s or %s (case sensitive)", REFRESH, REFETCH))),
+	)
+
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+
+	if job.StartTime == nil {
+		return errors.New(strings.Join(append(errs, "startTime: cannot be blank"), ", "))
+	}
+
+	if job.StartTime.After(time.Now().Add(twoDays)) {
+		errs = append(errs, "startTime: must be within two days from now")
+	}
+
+	if job.StartTime.Before(time.Now()) {
+		errs = append(errs, "startTime: cannot be in the past")
+	}
+
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, ", "))
+	}
+
+	return nil
+}
+
+// validateDeliveryService ensures the supplied (required) Delivery Service XML ID exists
+func (job *InvalidationJobCreateV40) validateDeliveryService(tx *sql.Tx) error {
+	var exists bool
+	row := tx.QueryRow(`SELECT EXISTS(SELECT * FROM deliveryservice WHERE xml_id=$1)`, job.DeliveryService)
+	if err := row.Scan(&exists); err != nil {
+		if err == sql.ErrNoRows {
+			return fmt.Errorf("No DeliveryService exists matching identifier: %v", job.DeliveryService)
+		}
+		return errors.New("Unknown error occurred")
+	}
+	return nil
+}
+
+// validateTLLHours ensures the supplied TTL hours is within acceptable limits
+func (job *InvalidationJobCreateV40) validateTLLHours(tx *sql.Tx) error {
+	var maxDays uint
+	err := tx.QueryRow(`SELECT value FROM parameter WHERE name='maxRevalDurationDays' AND config_file='regex_revalidate.config'`).Scan(&maxDays)
+	maxHours := maxDays * 24
+	if err == nil && uint(job.TTLHours) > maxHours { // silently ignore other errors too
+		return fmt.Errorf("cannot exceed %s", strconv.FormatUint(uint64(maxHours), 10))
+	}
+	return nil
+}
+
+func (job *InvalidationJobCreateV40) GetDSIDfromDSXMLID(tx *sql.Tx) (uint, error) {
+	var dsID uint
+	row := tx.QueryRow(`SELECT id FROM deliveryservice WHERE xml_id=$1`, job.DeliveryService)
+	if err := row.Scan(&dsID); err != nil {
+		if err == sql.ErrNoRows {
+			return 0, fmt.Errorf("No DeliveryService exists matching identifier: %v", job.DeliveryService)
+		}
+		return 0, errors.New("Unknown error occurred")
+	}
+	return dsID, nil
+}
+
+// RefetchAllowed checks whether Refetch is allowed and enabled as a global paramter
+// TODO: Error is swallowed from DB...
+func RefetchAllowed(tx *sql.Tx) bool {
+	refetchEnabled := false
+	// Silently ignore paramter error
+	tx.QueryRow(`SELECT value FROM parameter WHERE name='refetch_enabled' AND config_file='global'`).Scan(&refetchEnabled)
```

Comment:
I assume that this won't always fail - which is what I would have guessed, but I'm sure you'd have noticed that -, but it for sure can be made to fail. Parameter values are strings, not booleans, so even if this coerces values `'true'` and `'false'` to booleans, it will for sure fail if the value of the Parameter is something like `'foo'`, which means that you can make a Parameter that's perfectly valid that will break an API endpoint so that it only ever returns 500 responses.

We also definitely don't want to be swallowing that error.

Further, the problem with this is that there is no validation on Parameters, and that WHERE clause can even match multiple Parameters. Plus, by convention, we put global Parameters in the GLOBAL Profile - but that Profile could also be deleted or renamed at any time.

So if I have two Parameters:
```json
[{
    "name": "refetch_enabled",
    "configFile": "global",
    "value": "true",
    "secure": false,
    "id": 1
},
{
    "name": "refetch_enabled",
    "configFile": "global",
    "value": "false",
    "secure": false,
    "id": 2
}]
```
Then the one chosen by this method is undefined, even if one of them shows up when you look at the GLOBAL profile, the other still might get chosen.

For all these reasons, I tend to prefer using configuration file parameters (i.e. in "cdn.conf"), whose values can be exposed through the `system/info` endpoint if other components need to know about them.